gui/options.cpp

```diff
@@ -3564,4 +3606,306 @@ void GlobalOptionsDialog::storageErrorCallback(Networking::ErrorResponse respons
+	0xC86400, 0xE10000, 0x333333, 0x666666, 0x999999, 0xCCCCCC, 0xBFBFBF,
+	0xFE00FE,
+};
+
+static void boxColor(Graphics::Surface &surface, int x1, int y1, int x2, int y2, int color) {
+	surface.fillRect(Common::Rect(x1, y1, x2 + 1, y2 + 1), color);
+}
+
+static void squaremesh(Graphics::Surface &surface, int xres, int yres, int width, int height, int gapsize, int meshcolor, int squarecolor) {
+	int centerx = xres / 2;
+	int centery = yres / 2;
+
+	boxColor(surface, 0, 0, xres, yres, squarecolor);
+	for (int x = centerx % (width + gapsize) + width / 2 - (width + gapsize); x < xres; x += gapsize + width) {
+		boxColor(surface, x, 0, x + gapsize - 1, yres - 1, meshcolor);
+	}
+	for (int y = centery % (height + gapsize) + height / 2 - (height + gapsize); y < yres; y += gapsize + height) {
+		boxColor(surface, 0, y, xres - 1, y + gapsize - 1, meshcolor);
+	}
+}
+
+static void frame(Graphics::Surface &surface, int x1, int y1, int x2, int y2, int x3, int y3, int x4, int y4, int color) {
+	boxColor(surface, x1, y1, x2, y3, color);   //top
+	boxColor(surface, x1, y4, x2, y2, color);   //bottom
+	boxColor(surface, x1, y1, x3, y2, color);   //left
+	boxColor(surface, x4, y1, x2, y2, color);   //right
+}
+
+static void drawRow(Graphics::Surface &surface, int offsety, int colors[], int numColors, int gap, int vsize, int xres, int offsetx) {
+	int boxsize = (xres - (numColors + 1) * gap) / numColors;   //-gaps between boxes
+	offsetx += gap + 1;
+	for (int x = 0; x < numColors; x++) {
+		boxColor(surface, offsetx, offsety, offsetx + boxsize - 1, offsety + vsize - 1, colors[x]);
+		offsetx += (boxsize + gap);
+	}
+}
+
+static void circleColorNoblend(Graphics::Surface &surface, double x, double y, double r, int color, bool invert) {
+	int sx;
+
+	r += 0.49;
+	for (sx = (int)round(x - r); sx < (int)round(x + r) + 1; sx++) {
+		double h = sqrt(r * r - pow(sx - x, 2));
+		if (invert) {
+			surface.vLine(sx, 0, -1 + (int)ceil(y - h), color);
+			surface.vLine(sx, 1 + (int)floor(y + h), surface.h, color);
+		} else
+			surface.vLine(sx, (int)ceil(y - h), (int)floor(y + h), color);
+	}
+	if (invert && sx < surface.w)
+		boxColor(surface, sx, 0, surface.w - 1, surface.h - 1, color);
+	if (invert && (int)round(x - r) > 0)
+		boxColor(surface, 0, 0, -1 + (int)round(x - r), surface.h - 1, color);
+}
+
+static void blacksquareinwhitecircle(Graphics::Surface &surface, int x1, int y1, int gapsize, int rwidth, int rheight) {
+	double c1y = y1 - (gapsize + 1.0) / 2.0;
+	double c1x = x1 - (gapsize + 1.0) / 2.0;
+	circleColorNoblend(surface, c1x, c1y, rwidth * 1.5, WHITE, false);
+	boxColor(surface, x1 - rwidth - gapsize, y1 - rheight - gapsize, x1 + rwidth - 1, y1 + rheight - 1, BLACK);
+	boxColor(surface, x1 - rwidth - gapsize, y1 - gapsize, x1 + rwidth - 1, y1 - 1, WHITE);
+	boxColor(surface, x1 - gapsize, y1 - rheight - gapsize, x1 - 1, y1 + rheight - 1, WHITE);
+	circleColorNoblend(surface, c1x, c1y, gapsize + 1, WHITE, false);
+}
+
+static void drawTestPattern() {
```

Comment:
Would it be possible to have this in a separate file? It looks like it could be useful in the Testbed engine as well.

Reply:
Indeed. Moved to `graphics/pm5544.h`